### ISSUE TITLE
DRM/GBM support

### ISSUE THREAD

**Issue by icecream95** [ptitSeb/gl4es#123]
The LIBGL_FPS value seems capped at around 20 - for games which definitely feel closer to 60 fps (and can get around 20 fps just with llvmpipe, for low resolutions...), the FPS value from gl4es is usually at 21 or 22, sometimes going up to around 25 or so.

This, combined with the tendency of a certain scenegraph library to crash when bringing up the FPS display, makes it hard to get an accurate idea of what the framerate actually is.

**Comment by ptitSeb** [comment]
The FPS counter use `gettimeofday(...)`. Maybe the precision of that function is not that good on your system.
I can probably add a build option to use `clock_gettime(...)` instead maybe.

(on the machine I tests, I can have fps like 220fps, so not limited to 20).

**Comment by kas1e** [comment]
When i do test on amigaos4, i also didn't have 20 fps limit, it can show and 50 and 60 and 200

**Comment by ptitSeb** [comment]
@icecream95 I have puched the option, try to activate if (USE_CLOCK) and see if you have better fps counter with that.

**Comment by icecream95** [comment]
I tried running glxgears, and both gl4es's FPS counter and that of glxgears report only 45 fps.

So it seems the problem is not with gl4es reporting the wrong frame rate, but rather with fbdev having a large overhead with the mali blob driver.

glmark2-es2-drm can get around 60 FPS, though. It's only a couple hundred lines to get a GLES context with the DRM API, so it looks like it's time for me to add another option to LIBGL_FB for drm...

**Comment by icecream95** [comment]
What's the status for GBM support?

```
LIBGL: Initialising gl4es
LIBGL: v1.1.1 built on Sep  7 2019 18:03:28
LIBGL: using GBM
LIBGL: Using GLES 2.0 backend
LIBGL: loaded: libGLESv2.so
LIBGL: loaded: libEGL.so
LIBGL: loaded: libgbm.so
LIBGL: GBM on card /dev/dri/card0 is Available
LIBGL: Using GLES 2.0 backend
LIBGL: Error while gathering supported extension (eglInitialize: EGL_BAD_DISPLAY), default to none
LIBGL: Targeting OpenGL 2.0
LIBGL: Not trying to batch small subsequent glDrawXXXX
LIBGL: VBO used (in a few cases)
LIBGL: glX Will try to recycle EGL Surface
LIBGL: Current folder is:/
LIBGL: Warning, no EGLConfig matching GBM Format found
LIBGL: ERROR: EGL Error detected: EGL_BAD_DISPLAY
LIBGL: warning, egl_eglCreatePlatformWindowSurface is NULL
```

**Comment by ptitSeb** [comment]
Yeah, GBM is not ready yet. I still have have a few bugs in context creation that I have to fix.

Is the FPS counter fixed for you now ?

**Comment by icecream95** [comment]
The FPS counter was always working. I just didn't think that the overhead for framebuffer rendering could make the frame rate so low.

I get a smooth 20fps at least, so I'm not complaining that much (though I am still trying to get DRM/GBM/KMS working).

I just knew something was fishy about the bottleneck for OpenMW being rendering rather than draw calls...

**Comment by icecream95** [comment]
Over the past day I have been working on a DRM wrapper for EGL, and I've currently got es2gears working with it.

The repo is at <https://gitlab.com/icecream95/drmegl-wrapper>.

I will be busy over the next couple of weeks, but you are welcome to try to get it working with gl4es.

**Comment by ptitSeb** [comment]
Ah nice, thanks. I'll take a look and see if I can fix / change my implementation.

Also, did you have some speed improvment avec simple EGL/FrameBuffer driver (or maybe it's still difficult to say) ?

**Comment by icecream95** [comment]
I got around 115 fps with es2gears, which is a definite improvement over fbdev (45 fps for glxgears).

Note that the code I wrote doesn't do as much error checking as it should, and doesn't do any cleanup at the end (does that need to be explicitly done?).

Also, the drm surface needs to be released somehow on VT switch so it doesn't kill X.

For how to implement vsync, look [here](https://gist.github.com/Miouyouyou/89e9fe56a2c59bce7d4a18a858f389ef#file-linux_drm_opengles-c-L404).

**Comment by icecream95** [comment]
I got glxgears working through gl4es, and the FPS increase is pretty big:

```
232 frames in 5.0 seconds = 46.338 FPS
929 frames in 5.0 seconds = 185.706 FPS
```

All I did to get it to work was to only setup GBM once, and to just return from `eglMakeCurrent` without modesetting when the buffer object it gets from gbm is NULL, so that it doesn't attempt (and fail) to show PBuffers on screen.

I wonder what score I'll get in glmark...

**Comment by ptitSeb** [comment]
Nice speed increase !
I really need to fix my GBM code in gl4es.

**Comment by icecream95** [comment]
My wrapper seems (so far) to work fine, so you could just use that, possibly integrating it into the gl4es source, and then just drop the current GBM code.

glmark2 seems to ignore gl4es -- I bet it was compiled to use libglvnd...

**Comment by ptitSeb** [comment]
So, I have merge your wrapper in gl4es. I hope I haven't broke anything. I tested on my linux vm, but I got a black window, and got nothing conclusive from Ctrl-Alt-2 virtual terminal either. But I guess that could be expected from a VM (also, I tried your wrapper, but it dies with a `failed to set mode: Permission denied`  that I don't get with gl4es, strange).

**Comment by icecream95** [comment]
With systemd, modesetting shouldn't require any extra privileges when logging in from a VT.
However, adding yourself to the video group or running `chmod a+rw /dev/dri/card0` should make it work.
I recall some issues to do with having multiple cards in `/dev/dri` - if you have multiple cards in there you should try with each of them.

I haven't tried the merged code yet, but at least with my wrapper SDL2 programs don't work.
I know `eglMakeCurrent` is called with a NULL surface, but haven't done any further debugging yet.
Neverball is a good one to try, as it doesn't even make it to the `drmModeSetCrtc` call, so you can use gdb without doing anything special.

My wrapper seems to call back to gl4es functions, as it uses `RTLD_NEXT` with `dlsym`. I've committed a change (that I haven't tested yet...) that uses `dlopen` with a file path, but you'll need to change the `EGLLIB` define to the location of the `libEGL.so` on your system. I think the merged version probably doesn't have this problem.

**Comment by icecream95** [comment]
[Here](https://github.com/ptitSeb/gl4es/blob/eedc58d85b70f58e8f683333e93cd0f0e9902958/src/glx/gbm.c#L360), you return EGL_FALSE if the buffer object is NULL.
The buffer object is NULL when the surface is a PBuffer, so it is not an error condition.
Unless you were careful not to use `GBMMakeCurrent` for PBuffer surfaces, you should return EGL_TRUE.

I think the SDL2 problem could be solved by doing some more careful filtering in `eglCreateWindowSurface`, rather than just passing a NULL `attrib_list`.

**Comment by icecream95** [comment]
I've noticed that I get really bad tearing, which seems to be arranged in tiles.
Mobile GPUs do tile-based rendering, so I think what it is doing is swapping half-rendered frames to the screen.
At least that means it will have low latency :D

**Comment by icecream95** [comment]
I tested my wrapper on a x86_64 system (Intel graphics), and the only change I had to make was linking es2gears with EGL.

I got 3400 fps, but got short, flickering black lines were onscreen, in what I think is a similar issue to the tile-based tearing on Mali.

**Comment by ptitSeb** [comment]
I also had to add EGL to es2gears. I'll try to add my account to video group and try again. I'll also fix the "PBuffer" issue (but I don't think it will work for now in gl4es, as I specify the "PBUFFER" bit when creating the EGL Context and I think it must not be there).

**Comment by icecream95** [comment]
I tried making `eglSwapBuffers` only actually swap the buffers every tenth call.

The tearing only lasted for one frame, so it is definitely the GPU drawing to the FB after it was been swapped out to the display.

I'm pretty sure this is what is happening:

1. FB 1 is the front buffer, FB 2 is the back buffer
2. The GPU starts rendering to FB 2
3. eglSwapBuffers makes FB 2 the front buffer and FB 1 the new back buffer
4. The GPU finishes rendering on FB 2 (which is currently displayed!)
5. The GPU starts rendering the next frame on FB 1

I would probably say this was a driver issue, but it happens with Intel graphics too.

Any ideas on why this is happening, and how to fix it?

`glFinish` didn't seem to help much.

I'll see if vsync helps after I reboot - systemd is using up 100% CPU...

**Comment by ptitSeb** [comment]
So you are saying the GPU doesn't acknowledge the `glFinish()` command continue it's drawing in the "old" backbuffer?

**Comment by icecream95** [comment]
It turns out putting vsync and three calls to the driver eglSwapBuffers in the eglSwapBuffers code fixed the problems with the tearing.

Triple buffering should work too and be lower overhead, but vsync works well enough.

The main problem now is SDL2 not working.

**Comment by icecream95** [comment]
I compiled the 5.3 kernel and the Panfrost mesa driver, and now everything Just Works™ with Wayland, including GLX through XWayland, with the added bonus of not having ARM uploading my passwords any more, as they were probably doing with the blob driver.

That means that I'm not using DRM [directly] any more, and can just run programs on Wayland compositors normally. So, I won't continue doing anything with DRM, but I will still be happy to boot into the old kernel to fix any problems you (or anyone else) might be having.

If you want to continue working on DRM output, the main things that have to be done are:

- SDL2 support
- Hooking into VT switch to do `drmSetMaster` and `drmDropMaster`, to avoid killing X
- Triple buffering

If I can be bothered, I'll document how I got DRM and FB rendering working with Xdummy for input, but it's just so nice playing Supertuxkart without the framerate seemingly arbitrarily capped at 20 fps, so you might have to wait a while :)

Panfrost actually has native support for some desktop OpenGL functions, so I can actually run most programs without gl4es, but a few programs require gl4es with ES=1 - so far, everything that works with LIBGL_ES=2 also works natively, which I find interesting.

A few games (so far: OpenMW, GNUjump, Assaultcube, Warzone2100 and Extreme Tux Racer (which never worked with the blob driver anyway)) have bad flickering or random triangles covering half the screen which make them unplayable, which isn't related to gl4es, but everything else is working quite nicely so far.

TLDR: I'm closing this issue, as I don't use DRM directly any more. Reopen it if you want.
